aten/src/ATen/native/BatchLinearAlgebra.cpp

```diff
@@ -1569,6 +1569,361 @@ std::tuple<Tensor&, Tensor&> symeig_out(Tensor& vals, Tensor& vecs, const Tensor
+  }
+
+  // 'vectors' must be in batched column major order (Fortran contiguous)
+  if (compute_eigenvectors) {
+    TORCH_INTERNAL_ASSERT(vectors.transpose(-2, -1).is_contiguous());
+    TORCH_INTERNAL_ASSERT(vectors.sizes().equals(input.sizes()));
+  }
+
+  // 'values' must be contiguous
+  TORCH_INTERNAL_ASSERT(values.is_contiguous());
+  TORCH_INTERNAL_ASSERT(values.sizes().equals(values_shape));
+
+  // if 'input' is complex then use 'values' directly else create a temporary to hold the real and imaginary parts
+  // and then use at::complex_out
+  Tensor real_imag_values = values;
+
+  // if 'input' is complex then use 'vectors' directly else maybe create a temporary to hold real vectors
+  // and then use linalg_eig_make_complex_eigenvectors
+  Tensor maybe_complex_vectors = vectors;
+  if (!input.is_complex()) {
+    // first n elements to hold the real portion of the output and the last n elements to hold the imaginary portion
+    auto real_imag_shape = IntArrayRef(input.sizes().data(), input.dim()-2).vec();  // input.shape[:-2]
+    real_imag_shape.push_back(input.size(-1) * 2);
+    real_imag_values = at::empty(real_imag_shape, options, MemoryFormat::Contiguous);
+
+    // linalg_eig_stub expects real-valued tensor to store eigenvectors
+    // output of linalg_eig_stub need to be post-processed later to produce complex-valued eigenvectors
+    // we do this post-processing only if 'vectors' is complex-valued
+    // otherwise storage of 'vectors' is used directly
+    if (vectors.is_complex() && compute_eigenvectors) {
+      maybe_complex_vectors = at::empty(input.sizes(), options, MemoryFormat::Contiguous);
+      maybe_complex_vectors.transpose_(-2, -1);  // make 'maybe_complex_vectors' to have Fortran contiguous memory layout
+    }
+  }
+
+  // now call the actual computation
+  linalg_eig_stub(input.device().type(), real_imag_values, maybe_complex_vectors, infos, input, compute_eigenvectors);
+
+  // if input is not complex we need to do some post-processing
+  if (!input.is_complex()) {
+    // extract real and imaginary parts of the output
+    auto real_values = real_imag_values.slice(/*dim=*/-1, /*start=*/0, /*end*/input.size(-1));
+    auto imag_values = real_imag_values.slice(/*dim=*/-1, /*start=*/input.size(-1));
+
+    // if the imaginary part is zero we don't need to do anything
+    bool is_zero_imag = at::all(imag_values == 0.0).item().toBool();
+    if (is_zero_imag) {
+      values.copy_(real_values);
+      if (compute_eigenvectors) {
+        vectors.copy_(maybe_complex_vectors);  // does nothing for !vectors.is_complex() because vectors.is_same(maybe_complex_vectors) == true
+      }
+      return std::tuple<Tensor&, Tensor&>(values, vectors);
+    }
+
+    if (values.is_complex()) {
+      values = at::complex_out(values, real_values, imag_values);
+    } else {
+      TORCH_CHECK(false, "linalg_eig: imaginary part of eigenvalues is non-zero, can't safely cast eigenvalues to non-complex dtype.")
```

Comment:
"linalg_eig" -> "torch.linalg.eig()". Let's start preferring the user-facing names in the error messages.